### ISSUE TITLE
Reproducible TestDrillSideways failure

### ISSUE THREAD

**Issue by jpountz** [apache/lucene#12418]
### Description

The following gradle command fails reproducibly on `branch_9x` with the following error:

```
   >     java.lang.AssertionError
   >         at __randomizedtesting.SeedInfo.seed([EA93FFB2BD206B3C:98DFDABD0C40DD4F]:0)
   >         at junit@4.13.1/org.junit.Assert.fail(Assert.java:87)
   >         at junit@4.13.1/org.junit.Assert.assertTrue(Assert.java:42)
   >         at junit@4.13.1/org.junit.Assert.assertFalse(Assert.java:65)
   >         at junit@4.13.1/org.junit.Assert.assertFalse(Assert.java:75)
   >         at org.apache.lucene.facet.TestDrillSideways.verifyEquals(TestDrillSideways.java:1776)
   >         at org.apache.lucene.facet.TestDrillSideways.testRandom(TestDrillSideways.java:1338)
   >         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   >         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
   >         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   >         at java.base/java.lang.reflect.Method.invoke(Method.java:568)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at junit@4.13.1/org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
   >         at org.apache.lucene.test_framework@9.8.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
   >         at junit@4.13.1/org.junit.rules.RunRules.evaluate(RunRules.java:20)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
   >         at randomizedtesting.runner@2.8.1/com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
   >         at java.base/java.lang.Thread.run(Thread.java:833)
  2> NOTE: reproduce with: gradlew test --tests TestDrillSideways.testRandom -Dtests.seed=EA93FFB2BD206B3C -Dtests.locale=ar-OM -Dtests.timezone=Asia/Damascus -Dtests.asserts=true -Dtests.file.encoding=UTF-8
```

### Gradle command to reproduce

`./gradlew test --tests TestDrillSideways.testRandom -Dtests.seed=EA93FFB2BD206B3C -Dtests.locale=ar-OM -Dtests.timezone=Asia/Damascus -Dtests.asserts=true -Dtests.file.encoding=UTF-8`

**Comment by Yuti-G** [comment]
Hello Adrien, 

Thanks for raising this issue! I was trying to locate which release version / commit that broke this test in the beginning and I found that the test starts passing after I reverted this [commit](https://github.com/apache/lucene/commit/7d9de759f244896d47da77c842051845199d316e#diff-0a50ccaed94e831ddf0be77870286938e83620b3897379b198e450888fbac713). Since you were the original author, do you know why that would affect this drillsideways test? Thank you so much!

**Comment by jpountz** [comment]
Sorry @Yuti-G I had missed your reply!

I think that reverting the commit that you linked only made the test pass because this commit adds a call to `random().nextInt()` to `LuceneTestCase#newIndexWriterConfig()`. So by reverting this commit, `TestDrillSideways.testRandom()` got a difference sequence of random numbers for everything after the `newIndexWriterConfig()` call and the particular configuration that made this test fail was altered.

By the way if I revert the merge-on-refresh commit and then add back the call to `r.nextInt(3)`, the failure still reproduces:

diff
```
diff --cc lucene/core/src/test/org/apache/lucene/index/TestLogMergePolicy.java
index 0b00a20a17f,517542699c1..00000000000
--- a/lucene/core/src/test/org/apache/lucene/index/TestLogMergePolicy.java
+++ b/lucene/core/src/test/org/apache/lucene/index/TestLogMergePolicy.java
diff --git a/lucene/test-framework/src/java/org/apache/lucene/tests/util/LuceneTestCase.java b/lucene/test-framework/src/java/org/apache/lucene/tests/util/LuceneTestCase.java
index 1a5a876a7b7..d6649a26e9a 100644
--- a/lucene/test-framework/src/java/org/apache/lucene/tests/util/LuceneTestCase.java
+++ b/lucene/test-framework/src/java/org/apache/lucene/tests/util/LuceneTestCase.java
@@ -1037,6 +1037,8 @@ public abstract class LuceneTestCase extends Assert {
       c.setIndexWriterEventListener(new MockIndexWriterEventListener());
     }
 
+    r.nextInt(3);
+
     c.setMaxFullFlushMergeWaitMillis(rarely() ? atLeast(r, 1000) : atLeast(r, 200));
     return c;
   }

```

This looks unrelated to merge-on-refresh.

**Comment by benwtrent** [comment]
Another seed to reproduce:

```
./gradlew test --tests TestDrillSideways.testRandom -Dtests.seed=668BF9B25DA9EE8A
```

If I change the indexwriting config to:
```
IndexWriterConfig iwc = newIndexWriterConfig(new MockAnalyzer(random())).setMergePolicy(
        NoMergePolicy.INSTANCE);
```

The test passes.

**Comment by benwtrent** [comment]
I verified on all three of the failing seeds, setting `.setMergePolicy(NoMergePolicy.INSTANCE);` passes the test.

I don't know what that means and if we should consider that a valid fix for this or not.

Additionally, looking at the verbose output, some seeds have a `LogDocMergePolicy` and others a `TieredMergePolicy`. So, the policy kind doesn't seem to make a difference.

@jpountz

**Comment by gsmiller** [comment]
I think I know what's going on here, and I believe #12853 fixes it. In both seeds mentioned here, I've noticed that the drill-sideways query short-circuits on at least one segment, knowing it has no results. Specifically, a null BulkScorer gets returned from DrillSidewaysQuery. The problem with this is that the "sideways" facet collectors get left hanging, and nothing calls #finish on them.

I reproduced with both seeds on 9x and main, and verified the above fix works in all cases.
